File: python/base_request.py
# -*- coding: utf-8 -*-
"""
拼多多 API 请求基类（自包含版）
==============================
不依赖项目内 logger/db_manager/登录模块，完全自包含。
提供：重试机制、域名节流、Anti-Content 自动注入、cookies 管理。
"""
import json
import time
import random
import logging
import threading
import requests
from typing import Dict, Any, Optional, Union, Callable
from urllib.parse import urlparse

from anti_content import DEFAULT_USER_AGENT, generate_anti_content, requires_anti_content

logger = logging.getLogger("pdd_api")


class BaseRequest:
    """
    API 请求基类，统一管理 requests 请求。

    功能：
    - 指数退避重试 + 随机抖动
    - 按域名请求频率控制（mms.pinduoduo.com = 0.8s，其他 = 0.5s）
    - Anti-Content 头自动注入（白名单端点）
    - 统一错误处理与日志（敏感字段脱敏）

    注意：本类不实现自动重新登录。会话过期时返回响应数据，
    由上层（auth.py / 调用方）决定是否重新登录。
    """

    SESSION_EXPIRED_ERROR_CODE = 43001
    RETRY_JITTER_MIN = 0.1
    RETRY_JITTER_MAX = 0.3

    _last_request_time: Dict[str, float] = {}
    _rate_lock = threading.Lock()
    _DOMAIN_INTERVALS = {"mms.pinduoduo.com": 0.8}
    _DEFAULT_MIN_INTERVAL = 0.5

    _SENSITIVE_KEYS = {'password', 'cookies', 'token', 'api_key',
                       'access_token', 'anti-content', 'anti_content'}

    def __init__(self, cookies: Optional[Dict] = None, mall_id: Optional[str] = None,
                 max_retries: int = 3, retry_delay: float = 1.0, retry_backoff: float = 2.0,
                 min_request_interval: float = 0.5, auto_relogin: bool = False,
                 relogin_callback: Optional[Callable] = None):
        self.cookies = cookies or {}
        self.mall_id = mall_id
        self.max_retries = max_retries
        self.retry_delay = retry_delay
        self.retry_backoff = retry_backoff
        self.min_request_interval = min_request_interval
        self.auto_relogin = auto_relogin
        self._relogin_callback = relogin_callback
        self._relogin_attempted = False

        self.default_headers = {
            'accept': 'application/json, text/plain, */*',
            'accept-language': 'zh-CN,zh;q=0.9,en;q=0.8',
            'content-type': 'application/json;charset=UTF-8',
            'origin': 'https://mms.pinduoduo.com',
            'user-agent': DEFAULT_USER_AGENT,
            'priority': 'u=1, i',
        }

    # ── 请求头构建 ──────────────────────────────────────────────────────

    def _build_headers(self, url: str = "", referer: Optional[str] = None,
                       extra: Optional[Dict[str, str]] = None) -> Dict[str, str]:
        """构建请求头，自动注入 anti-content（仅白名单端点）"""
        headers = self.default_headers.copy()
        if referer:
            headers['referer'] = referer
        else:
            headers['referer'] = 'https://mms.pinduoduo.com/'
        # anti-content 自动注入
        if url and requires_anti_content(url):
            try:
                ac = generate_anti_content()
                if ac:
                    headers['anti-content'] = ac
            except Exception as e:
                logger.warning("生成 anti-content 失败（端点 %s）: %s", url, e)
        if extra:
            headers.update(extra)
        return headers

    # ── 频率控制 ──────────────────────────────────────────────────────

    def _rate_limit(self, url: str) -> None:
        """按域名节流，避免高频触发风控"""
        if not self.min_request_interval or not url:
            return
        try:
            domain = urlparse(url).hostname or ""
        except Exception:
            return
        if not domain:
            return
        interval = self._DOMAIN_INTERVALS.get(domain, self.min_request_interval)
        if interval <= 0:
            return
        with self._rate_lock:
            now = time.time()
            last = self._last_request_time.get(domain, 0.0)
            wait = interval - (now - last)
            if wait > 0:
                logger.debug("频率控制: %s 等待 %.3fs", domain, wait)
                time.sleep(wait)
            self._last_request_time[domain] = time.time()

    # ── 重试机制 ──────────────────────────────────────────────────────

    def _should_retry(self, response: Optional[requests.Response] = None,
                      exception: Optional[Exception] = None) -> bool:
        if exception:
            if isinstance(exception, (requests.ConnectionError, requests.Timeout,
                                      requests.HTTPError, requests.TooManyRedirects)):
                return True
        if response:
            if response.status_code >= 500:
                return True
            if response.status_code in (429, 408, 502, 503, 504):
                return True
        return False

    def _calc_retry_delay(self, attempt: int) -> float:
        delay = self.retry_delay * (self.retry_backoff ** attempt)
        jitter = random.uniform(self.RETRY_JITTER_MIN, self.RETRY_JITTER_MAX) * delay
        return delay + jitter

    def _is_session_expired(self, data: Dict[str, Any]) -> bool:
        if not data:
            return False
        if (data.get('error_code') == self.SESSION_EXPIRED_ERROR_CODE and
                '会话已过期' in str(data.get('error_msg', ''))):
            return True
        return False

    def _execute_with_retry(self, request_func: Callable, expect_json: bool = True,
                            url: str = "") -> Optional[Dict[str, Any]]:
        last_exception = None
        last_response = None
        for attempt in range(self.max_retries + 1):
            try:
                if attempt == 0 and url:
                    self._rate_limit(url)
                response = request_func()
                if response and response.status_code == 200:
                    data = self._handle_response(response, expect_json)
                    # 会话过期处理
                    if (data and self._is_session_expired(data) and
                            not self._relogin_attempted and self.auto_relogin and
                            self._relogin_callback):
                        logger.info("检测到会话过期，尝试重新登录...")
                        self._relogin_attempted = True
                        new_cookies = self._relogin_callback(self.mall_id)
                        if new_cookies:
                            self.update_cookies(new_cookies)
                            continue
                    return data
                last_response = response
                if attempt < self.max_retries and self._should_retry(response=response):
                    delay = self._calc_retry_delay(attempt)
                    logger.warning("请求失败 status=%s，第 %d 次重试，延迟 %.2fs",
                                   response.status_code, attempt + 1, delay)
                    time.sleep(delay)
                    continue
                return self._handle_response(response, expect_json)
            except Exception as e:
                last_exception = e
                if attempt < self.max_retries and self._should_retry(exception=e):
                    delay = self._calc_retry_delay(attempt)
                    logger.warning("请求异常: %s，第 %d 次重试，延迟 %.2fs",
                                   e, attempt + 1, delay)
                    time.sleep(delay)
                    continue
                logger.error("请求最终失败: %s", e)
                return None
        if last_exception:
            logger.error("重试 %d 次后仍失败: %s", self.max_retries, last_exception)
        elif last_response:
            logger.error("重试 %d 次后仍失败，最后状态码: %s",
                         self.max_retries, last_response.status_code)
        return None

    # ── 响应处理 ──────────────────────────────────────────────────────

    def _handle_response(self, response: requests.Response,
                         expect_json: bool = True) -> Optional[Dict[str, Any]]:
        try:
            if response.status_code != 200:
                logger.error("请求失败 status=%s, body=%s",
                             response.status_code, response.text[:500])
                return None
            if expect_json:
                try:
                    return response.json()
                except json.JSONDecodeError:
                    logger.error("JSON 解析失败: %s", response.text[:500])
                    return None
            return {"text": response.text, "status_code": response.status_code}
        except Exception as e:
            logger.error("处理响应出错: %s", e)
            return None

    def _sanitize(self, data: Any) -> Any:
        if not isinstance(data, dict):
            return data
        result = {}
        for k, v in data.items():
            if k.lower() in self._SENSITIVE_KEYS or any(s in k.lower() for s in self._SENSITIVE_KEYS):
                result[k] = '***'
            elif isinstance(v, dict):
                result[k] = self._sanitize(v)
            elif isinstance(v, list):
                result[k] = [self._sanitize(i) if isinstance(i, dict) else i for i in v]
            else:
                result[k] = v
        return result

    def _log_request(self, method: str, url: str, **kwargs):
        logger.debug("%s %s", method, url)
        params = kwargs.get('data') or kwargs.get('json')
        if params:
            logger.debug("请求参数: %s", self._sanitize(params))

    # ── 公开请求方法 ──────────────────────────────────────────────────

    def get(self, url: str, params: Optional[Dict] = None,
            referer: Optional[str] = None, headers: Optional[Dict[str, str]] = None,
            timeout: int = 30, expect_json: bool = True, **kwargs) -> Optional[Dict[str, Any]]:
        merged = self._build_headers(url, referer, headers)
        self._log_request("GET", url, params=params)

        def _make():
            return requests.get(url, params=params, headers=merged,
                                cookies=self.cookies, timeout=timeout, **kwargs)

        return self._execute_with_retry(_make, expect_json=expect_json, url=url)

    def post(self, url: str, data: Optional[Union[Dict, str]] = None,
             json_data: Optional[Dict] = None, referer: Optional[str] = None,
             headers: Optional[Dict[str, str]] = None, timeout: int = 30,
             expect_json: bool = True, **kwargs) -> Optional[Dict[str, Any]]:
        merged = self._build_headers(url, referer, headers)
        self._log_request("POST", url, data=data, json=json_data)

        def _make():
            return requests.post(url, data=data, json=json_data, headers=merged,
                                 cookies=self.cookies, timeout=timeout, **kwargs)

        return self._execute_with_retry(_make, expect_json=expect_json, url=url)

    # ── 工具方法 ──────────────────────────────────────────────────────

    def generate_request_id(self) -> int:
        return int(time.time() * 1000)

    def update_cookies(self, new_cookies: Union[Dict, str]) -> None:
        if isinstance(new_cookies, str):
            try:
                self.cookies = json.loads(new_cookies)
            except json.JSONDecodeError:
                logger.error("更新 cookies 失败: JSON 解析错误")
        elif isinstance(new_cookies, dict):
            self.cookies = new_cookies
        else:
            logger.error("更新 cookies 失败: 不支持的数据类型")

    def set_default_header(self, key: str, value: str) -> None:
        self.default_headers[key] = value

    def set_retry_config(self, max_retries: int = None, retry_delay: float = None,
                         retry_backoff: float = None) -> None:
        if max_retries is not None:
            self.max_retries = max_retries
        if retry_delay is not None:
            self.retry_delay = retry_delay
        if retry_backoff is not None:
            self.retry_backoff = retry_backoff

File: nodejs/base_request.js
'use strict';
/**
 * 拼多多 API 请求基类（Node.js 自包含版）
 * ============================================
 * 零第三方依赖，仅用 Node.js 内置 https/http 模块。
 * 提供：重试机制、域名节流、Anti-Content 自动注入、cookies 管理。
 */
const https = require('https');
const http = require('http');
const { URL } = require('url');
const { generateAntiContent, requiresAntiContent, DEFAULT_USER_AGENT } = require('./anti_content');

// ── 全局频率控制（按域名） ──────────────────────────────────────────
const _lastRequestTime = {};
const _DOMAIN_INTERVALS = { 'mms.pinduoduo.com': 0.8 };
const _DEFAULT_MIN_INTERVAL = 0.5;
const _SENSITIVE_KEYS = ['password', 'cookies', 'token', 'api_key', 'access_token', 'anti-content', 'anti_content'];

function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

function sanitize(data) {
    if (!data || typeof data !== 'object') return data;
    const result = {};
    const sensitiveSet = new Set(_SENSITIVE_KEYS.map(s => s.toLowerCase()));
    for (const k of Object.keys(data)) {
        const lk = k.toLowerCase();
        if (sensitiveSet.has(lk)) {
            result[k] = '***';
        } else if (typeof data[k] === 'object' && data[k] !== null) {
            result[k] = Array.isArray(data[k]) ? data[k].map(sanitize) : sanitize(data[k]);
        } else {
            result[k] = data[k];
        }
    }
    return result;
}

class BaseRequest {
    /**
     * @param {Object} opts
     * @param {Object} [opts.cookies] - cookies 字典
     * @param {string} [opts.mallId] - 店铺 ID
     * @param {number} [opts.maxRetries=3]
     * @param {number} [opts.minInterval=0.5] - 最小请求间隔（秒）
     * @param {boolean} [opts.autoRelogin=false]
     * @param {Function} [opts.reloginCallback]
     */
    constructor(opts = {}) {
        this.cookies = opts.cookies || {};
        this.mallId = opts.mallId || null;
        this.maxRetries = opts.maxRetries != null ? opts.maxRetries : 3;
        this.retryDelay = opts.retryDelay || 1.0;
        this.retryBackoff = opts.retryBackoff || 2.0;
        this.minInterval = opts.minInterval != null ? opts.minInterval : 0.5;
        this.autoRelogin = !!opts.autoRelogin;
        this._reloginCallback = opts.reloginCallback || null;
        this._reloginAttempted = false;

        this.defaultHeaders = {
            'accept': 'application/json, text/plain, */*',
            'accept-language': 'zh-CN,zh;q=0.9,en;q=0.8',
            'content-type': 'application/json;charset=UTF-8',
            'origin': 'https://mms.pinduoduo.com',
            'user-agent': DEFAULT_USER_AGENT,
            'priority': 'u=1, i',
        };
    }

    // ── 请求头构建 ──────────────────────────────────────────────────

    _buildHeaders(url, referer, extra) {
        const headers = Object.assign({}, this.defaultHeaders);
        headers['referer'] = referer || 'https://mms.pinduoduo.com/';
        if (url && requiresAntiContent(url)) {
            try {
                const ac = generateAntiContent();
                if (ac) headers['anti-content'] = ac;
            } catch (e) {
                console.warn('[anti_content] 生成失败:', e.message);
            }
        }
        if (this.cookies && Object.keys(this.cookies).length) {
            headers['cookie'] = Object.entries(this.cookies).map(([k, v]) => `${k}=${v}`).join('; ');
        }
        if (extra) Object.assign(headers, extra);
        return headers;
    }

    // ── 频率控制 ────────────────────────────────────────────────────

    async _rateLimit(url) {
        if (!this.minInterval || !url) return;
        let domain;
        try { domain = new URL(url).hostname || ''; } catch (e) { return; }
        if (!domain) return;
        const interval = _DOMAIN_INTERVALS[domain] != null ? _DOMAIN_INTERVALS[domain] : this.minInterval;
        if (interval <= 0) return;
        const now = Date.now();
        const last = _lastRequestTime[domain] || 0;
        const wait = interval * 1000 - (now - last);
        if (wait > 0) await sleep(wait);
        _lastRequestTime[domain] = Date.now();
    }

    // ── 底层请求 ────────────────────────────────────────────────────

    _rawRequest(method, url, { params, body, referer, headers, timeout = 30000, _redirectCount = 0 } = {}) {
        const MAX_REDIRECTS = 5;
        return new Promise((resolve, reject) => {
            let fullUrl = url;
            if (params && Object.keys(params).length) {
                const qs = new URLSearchParams();
                for (const [k, v] of Object.entries(params)) qs.append(k, String(v));
                fullUrl += (url.includes('?') ? '&' : '?') + qs.toString();
            }
            const u = new URL(fullUrl);
            const lib = u.protocol === 'https:' ? https : http;
            const mergedHeaders = this._buildHeaders(url, referer, headers);
            let payload = null;
            if (body != null) {
                payload = typeof body === 'string' ? body : JSON.stringify(body);
                mergedHeaders['content-length'] = Buffer.byteLength(payload);
            }
            const reqOpts = {
                method,
                hostname: u.hostname,
                port: u.port || (u.protocol === 'https:' ? 443 : 80),
                path: u.pathname + u.search,
                headers: mergedHeaders,
                timeout,
            };
            const req = lib.request(reqOpts, (res) => {
                const redirectCodes = [301, 302, 307, 308];
                if (redirectCodes.includes(res.statusCode) && _redirectCount < MAX_REDIRECTS) {
                    const location = res.headers['location'];
                    if (location) {
                        const nextUrl = new URL(location, fullUrl).href;
                        const nextMethod = (res.statusCode === 307 || res.statusCode === 308) ? method : 'GET';
                        const nextBody = nextMethod === 'GET' || nextMethod === 'HEAD' ? null : body;
                        this._rawRequest(nextMethod, nextUrl, { body: nextBody, referer, headers, timeout, _redirectCount: _redirectCount + 1 })
                            .then(resolve, reject);
                        return;
                    }
                }
                let chunks = '';
                res.on('data', d => chunks += d);
                res.on('end', () => {
                    resolve({ status_code: res.statusCode, text: chunks, headers: res.headers });
                });
            });
            req.on('error', reject);
            req.on('timeout', () => { req.destroy(new Error('请求超时')); });
            if (payload) req.write(payload);
            req.end();
        });
    }

    // ── 重试机制 ────────────────────────────────────────────────────

    _shouldRetry(statusCode, exception) {
        if (exception) return true;
        if (statusCode >= 500) return true;
        if ([429, 408, 502, 503, 504].includes(statusCode)) return true;
        return false;
    }

    _calcRetryDelay(attempt) {
        const base = this.retryDelay * Math.pow(this.retryBackoff, attempt);
        const jitter = (0.1 + Math.random() * 0.2) * base;
        return (base + jitter) * 1000;
    }

    _isSessionExpired(data) {
        if (!data) return false;
        return data.error_code === 43001 && String(data.error_msg || '').includes('会话已过期');
    }

    async _executeWithRetry(requestFn, url) {
        let lastErr = null;
        let lastResp = null;
        for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
            try {
                if (attempt === 0 && url) await this._rateLimit(url);
                const resp = await requestFn();
                if (resp.status_code === 200) {
                    let data = null;
                    try { data = JSON.parse(resp.text); } catch (e) {
                        console.error('[JSON] 解析失败:', resp.text.slice(0, 300));
                        return null;
                    }
                    if (this._isSessionExpired(data) && !this._reloginAttempted &&
                        this.autoRelogin && this._reloginCallback) {
                        console.info('[session] 检测到会话过期，尝试重新登录...');
                        this._reloginAttempted = true;
                        const newCookies = await this._reloginCallback(this.mallId);
                        if (newCookies) { this.updateCookies(newCookies); continue; }
                    }
                    return data;
                }
                lastResp = resp;
                if (attempt < this.maxRetries && this._shouldRetry(resp.status_code)) {
                    const delay = this._calcRetryDelay(attempt);
                    console.warn(`[retry] status=${resp.status_code}，第 ${attempt + 1} 次重试，延迟 ${(delay / 1000).toFixed(2)}s`);
                    await sleep(delay);
                    continue;
                }
                console.error(`[fail] status=${resp.status_code}: ${resp.text.slice(0, 300)}`);
                try { return JSON.parse(resp.text); } catch (e) { return null; }
            } catch (e) {
                lastErr = e;
                if (attempt < this.maxRetries && this._shouldRetry(null, e)) {
                    const delay = this._calcRetryDelay(attempt);
                    console.warn(`[retry] 异常: ${e.message}，第 ${attempt + 1} 次重试`);
                    await sleep(delay);
                    continue;
                }
                console.error(`[fail] 请求最终失败: ${e.message}`);
                return null;
            }
        }
        if (lastErr) console.error(`[retry] ${this.maxRetries} 次后仍失败: ${lastErr.message}`);
        else if (lastResp) console.error(`[retry] ${this.maxRetries} 次后仍失败，status=${lastResp.status_code}`);
        return null;
    }

    // ── 公开请求方法 ────────────────────────────────────────────────

    async get(url, { params, referer, headers, timeout } = {}) {
        return this._executeWithRetry(() => this._rawRequest('GET', url, { params, referer, headers, timeout }), url);
    }

    async post(url, { json, data, referer, headers, timeout } = {}) {
        const body = json != null ? json : (data != null ? data : '');
        return this._executeWithRetry(() => this._rawRequest('POST', url, { body, referer, headers, timeout }), url);
    }

    // ── 工具方法 ────────────────────────────────────────────────────

    generateRequestId() {
        return Date.now();
    }

    updateCookies(newCookies) {
        if (newCookies && typeof newCookies === 'object') {
            this.cookies = newCookies;
        } else {
            console.error('[cookies] 更新失败：不支持的数据类型');
        }
    }
}

module.exports = { BaseRequest, DEFAULT_USER_AGENT, sanitize };
